aptos-move/aptos-intent/src/codegen.rs

```diff
@@ -0,0 +1,489 @@
+                        });
+                    }
+                },
+                BatchArgument::Signer(i) => {
+                    self.script.code.code.push(Bytecode::CopyLoc(*i as u8));
+                },
+                BatchArgument::Raw(_) => {
+                    let type_ = &self.script.signature_at(func_handle.parameters).0[idx];
+                    let inst_ty = if call.ty_args.is_empty() {
+                        type_.clone()
+                    } else {
+                        instantiate(type_, &subst_mapping)
+                    };
+                    let param_idx = self.parameters.len() as u8;
+                    self.parameters.push(inst_ty);
+                    self.script.code.code.push(Bytecode::MoveLoc(param_idx));
+                },
+            }
+        }
+
+        // Make function call
+        if call.ty_args.is_empty() {
+            self.script.code.code.push(Bytecode::Call(func_id));
+        } else {
+            if self.script.function_instantiations.len() >= TableIndex::MAX as usize {
+                return Err(PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS));
+            }
+            let fi_idx =
+                FunctionInstantiationIndex(self.script.function_instantiations.len() as u16);
+            let inst_sig = subst_mapping
+                .values()
+                .map(|idx| SignatureToken::TypeParameter(*idx))
+                .collect();
+
+            let type_parameters = self.add_signature(Signature(inst_sig))?;
+            self.script
+                .function_instantiations
+                .push(FunctionInstantiation {
+                    handle: func_id,
+                    type_parameters,
+                });
+            self.script.code.code.push(Bytecode::CallGeneric(fi_idx));
+        }
+
+        self.returned_val_to_local.push(self.locals.len() as u16);
+        self.return_counts
+            .push(self.script.signature_at(func_handle.return_).0.len() as u16);
+
+        // Instruction for storing return values
+        let ret_locals = self
+            .script
+            .signature_at(func_handle.return_)
+            .0
+            .iter()
+            .map(|ret| {
+                if call.ty_args.is_empty() {
+                    ret.clone()
+                } else {
+                    instantiate(ret, &subst_mapping)
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for ret_ty in ret_locals {
+            let local_idx = self.locals.len() + self.args.len() + self.signer_counts as usize;
+            if local_idx >= u8::MAX as usize {
+                return Err(PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS));
+            }
+            self.locals.push(ret_ty);
+            self.script.code.code.push(Bytecode::StLoc(local_idx as u8));
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Script {
```

Comment:
Same, feel like it might be helpful to separate out for `CompiledScript` building and (Aptos) transaction script generation. The added modularity will benefit other use cases that requires bytecode manipulation.